src/Microsoft.AspNet.Mvc.Razor/Precompilation/RazorPreCompileModule.cs

```diff
@@ -0,0 +1,101 @@
+        /// Instantiates a new <see cref="RazorPreCompileModule"/> instance.
+        /// </summary>
+        /// <param name="services">The <see cref="IServiceProvider"/> for the application.</param>
+        public RazorPreCompileModule(IServiceProvider services)
+        {
+            _loadContext = services.GetRequiredService<IAssemblyLoadContext>();
+        }
+
+        /// <summary>
+        /// Gets or sets a value that determines if symbols (.pdb) file for the precompiled views is generated.
+        /// </summary>
+        public bool GenerateSymbols { get; protected set; }
+
+        /// <inheritdoc />
+        /// <remarks>Pre-compiles all Razor views in the application.</remarks>
+        public virtual void BeforeCompile(BeforeCompileContext context)
+        {
+            var fileProvider = new PhysicalFileProvider(context.ProjectContext.ProjectDirectory);
+
+            MemoryCache memoryCache;
+            lock (_memoryCacheLookupLock)
+            {
+                var cacheKey = new PrecompilationCacheKey
+                {
+                    Configuration = context.ProjectContext.Configuration,
+                    TargetFramework = context.ProjectContext.TargetFramework
+                };
+
+                if (!_memoryCacheLookup.TryGetValue(cacheKey, out memoryCache))
+                {
+                    // When CompactOnMemoryPressure is true, the MemoryCache evicts items at every gen2 collection.
+                    // In DTH, gen2 happens frequently enough to make it undesirable for caching precompilation results. We'll
+                    // disable listening for memory pressure for the MemoryCache instance used by precompilation.
+                    memoryCache = new MemoryCache(new MemoryCacheOptions { CompactOnMemoryPressure = false });
+
+                    _memoryCacheLookup[cacheKey] = memoryCache;
+                }
+            }
+
+            var viewCompiler = new RazorPreCompiler(
+                context,
+                _loadContext,
+                fileProvider,
+                memoryCache)
+            {
+                GenerateSymbols = GenerateSymbols
+            };
+
+            viewCompiler.CompileViews();
+        }
+
+        /// <inheritdoc />
+        public void AfterCompile(AfterCompileContext context)
+        {
+        }
+
+        private class PrecompilationCacheKey : IEquatable<PrecompilationCacheKey>
+        {
+            public string Configuration { get; set; }
+
+            public FrameworkName TargetFramework { get; set; }
+
+            public bool Equals(PrecompilationCacheKey other)
+            {
+                return
+                    other.TargetFramework == TargetFramework &&
+                    string.Equals(other.Configuration, Configuration, StringComparison.Ordinal);
+            }
+
+            public override int GetHashCode()
+            {
+                return 13 * Configuration.GetHashCode() + TargetFramework.GetHashCode();
```

Comment:
HashCodeCombiner? Now in the handy-dandy `Microsoft.Framework.HashCodeCombiner.Sources` package :)